---
title: "Progression to dementia GWAS against eQTLgene and PsychEncode datasets"
author: "AlejandroMC, Regina Reynolds"
date: "15/12/2021"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: yes
---

```{r}
getwd()
```



```{r results="hide", warning=FALSE, message=FALSE}
.libPaths("/path/to/R_libs/")
library(colochelpR)
library(data.table)
library(MafDb.1Kgenomes.phase3.hs37d5)
library(tidyverse)
library(coloc)
library(biomaRt)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(DT)
library(R.utils)
```

```{r setup, include=FALSE}
theme_rhr <-  theme_bw(base_family = "Helvetica") +
  theme(panel.grid.major.x = element_blank(),
        legend.position = "right",
        strip.text = element_text(size = 8),
        strip.text.y = element_text(angle = 90),
        axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 8),
        axis.title.y = element_text(vjust = 0.6),
        axis.title = element_text(size = 10),
        panel.spacing = unit(0.1, "lines"))
knitr::opts_chunk$set(echo = T, warning = F, message= F)
```


# Description

We want to study if there is any variant among the progression to dementia GWAS hits that regulates the expression of any gene by using eQTLgene and PsychEncode eQTL data.
To do so, we are going to perform a colocalization study using the coloc.ABF Bayesian approach that allows us to know whether there is any shared causal genetic variant between our quantitative trait GWAS and the preselected eQTL datasets. In addition, we will use the colochelpR package as a wrapper to do coloc.
Of note, in colocalization analysis, we are not interested in knowing which is the causal variant, but whether there is an overlap between a given GWAS trait and any QTL df.
We can accomplish this by testing the H4 null hypothesis that says is there is/are any SNP associated with both the progression to dementia GWAS and the regulation in expression of any gene. In other words, we want to know if any GWAS hit or set of GWAS hits are associated with progression to dementia quantitative phenotype through the disregulation on expression / alternative splicing of any gene.

# Data load

# Loading GWAS
```{r eval=FALSE}
survDementia = fread("/path/to/metaanalysis_dementia.txt.gz") #File used to upload into FUMA containg the columns: CHR, BP, rsID, non_effect, effect, pval, Effect, StdErr, HR, lower95, upper95 
survDementia_raw_maf = fread("/path/to/metaanalysis_summary_statistics.tbl") #Output of METAL

```


# Getting all the info needed to perform coloc, and do some wrangling

```{r eval = FALSE}
# Extract maf, reformat
survDementia <- survDementia %>% 
  as_tibble() %>%
  inner_join(survDementia_raw_maf %>% dplyr::select(MarkerName, Allele1, Allele2, Freq1),
             by = c("rsID"="MarkerName", "non_effect"="Allele1", "effect"="Allele2")) %>%
  dplyr::rename(beta=Effect, p.value=pval, se=StdErr, maf = Freq1,
                A1 = non_effect, A2 = effect, chrpos = rsID)

# Get rsID to match SNP names from QTL dfs
dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37
survDementia_rsID_vbeta = colochelpR::convert_loc_to_rs(survDementia, dbSNP=dbSNP) %>%
  get_varbeta() %>%
  mutate(GWAS = "progToDementia")
head(as.data.frame(survDementia_rsID_vbeta))

# Deal with biallelic SNPs
survDementia_rsID_vbeta <- survDementia_rsID_vbeta %>%
  dplyr::filter(!(duplicated(SNP) | duplicated(SNP, fromLast = TRUE))) %>% 
  dplyr::bind_rows(survDementia_rsID_vbeta %>% 
                     dplyr::filter(duplicated(SNP) | duplicated(SNP, fromLast = TRUE)) %>%
                     dplyr::group_by(SNP) %>% 
                     dplyr::top_n(1, maf))


# save
fwrite(survDementia_rsID_vbeta, "harmonised_progtoDemGWAS.txt", quote = F, sep= "\t", col.names=T, row.names=F)
```


# Getting the genes in +-1MB window

We need to extract all the genes within +/- 1 Mb of all significant hits in the progression to dementia GWAS.

```{r eval = FALSE}
ensembl_gene_ids_overlapping_1Mb_window_hit <- survDementia_rsID_vbeta %>%
  colochelpR::get_genes_within_1Mb_of_signif_SNPs(pvalue_column="p.value",
                                                  CHR_column="CHR",
                                                  BP_column="BP", 
                                                  mart=37)

cat("total genes selected:", length(ensembl_gene_ids_overlapping_1Mb_window_hit), "\n")
```



```{r results="hide", warning = FALSE, message = FALSE, eval = FALSE}
results_path = "path/to/RESULTS"
results_path_GWAS_eQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "progtoDementia_eQTLgene")
# Path to the GWAS we load
GWAS_path <- "path/to/harmonised_progtoDemGWAS.txt"
eQTLgene_path <- "path/to/Full_eQTLgen_rsID.MAF.txt"
eQTLgene <- fread(eQTLgene_path)

df1_N = as.numeric(unique(survDementia_rsID_vbeta$TotalSampleSize))

for (index in seq_along(ensembl_gene_ids_overlapping_1Mb_window_hit)) {
  ensembl_geneID <- ensembl_gene_ids_overlapping_1Mb_window_hit[index]
  print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
  ciseQTL_filtered <- eQTLgene %>%
    dplyr::filter(Gene == ensembl_geneID) %>%
    check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
    dplyr::filter(!duplicated(SNP))
  
  df2_N = mean(ciseQTL_filtered$NrSamples)
 
  # Subset from which I Calculate the proportion cases mean.
  #df_subset <- survDementia_rsID_vbeta %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP) 
  
  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = survDementia_rsID_vbeta,
                                                           df2 = ciseQTL_filtered, 
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval", 
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2", 
                                                                     "Gene_2"),
                                                        df_1_name = "GWAS", 
                                                        df_2_name = "eQTL",
                                                        df1_path = GWAS_path,
                                                        df2_path=eQTLgene_path,
                                                            p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated, 
                                 results_dir_path=results_path_GWAS_eQTL)

  }
```



```{r results="hide", warning = FALSE, message = FALSE, eval = FALSE}
results_path = "path/to/RESULTS"
results_path_GWAS_eQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "progtoDementia_psychEncode")
# Path to the GWAS we load
GWAS_path <- "path/to/harmonised_progtoDemGWAS.txt"
psychEnc_path <- "path/to/PsychEncode/Full_hg19_cis-eQTL_rsID_MAF.txt"

psychEncode <- fread(psychEnc_path)

df1_N = as.numeric(unique(survDementia_rsID_vbeta$TotalSampleSize))

for (index in seq_along(ensembl_gene_ids_overlapping_1Mb_window_hit)) {
  ensembl_geneID <- ensembl_gene_ids_overlapping_1Mb_window_hit[index]
  print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
  ciseQTL_filtered <- psychEncode %>%
    dplyr::filter(Gene == ensembl_geneID) %>%
    check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
    dplyr::filter(!duplicated(SNP))
  
  if (isEmpty(ciseQTL_filtered$SNP)) {
    next
  }
  df2_N = mean(ciseQTL_filtered$number_of_SNPs_tested, na.rm = T)
  
 
  # Subset from which I Calculate the proportion cases mean.
  #df_subset <- survDementia_rsID_vbeta %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP) 
  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = survDementia_rsID_vbeta,
                                                           df2 = ciseQTL_filtered, 
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval", 
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2", 
                                                                     "Gene_2"),
                                                        df_1_name = "GWAS", 
                                                        df_2_name = "eQTL",
                                                        df1_path = GWAS_path,
                                                        df2_path=psychEnc_path,
                                                            p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated, 
                                 results_dir_path=results_path_GWAS_eQTL)
}

```




# Results 

Coloc calculates the posterior probability (PP) for 5 different hypotheses:
- H0: No association with either trait.
- H1: Association with trait 1, not with trait 2.
- H2: Association with trait 2, not with trait 1.
- H3: Association with trait 1 and 2, two independent SNPs.
- H4: Association with trait 1 and trait 2, one shared SNP.

We will set up a PP.H4 > 0.9 to consider the presence of a shared SNP between traits


```{r}
results_eQTL <- merge_coloc_summaries("path/to/progtoDementia_eQTLgene",
                                    add_signif_SNP = F, recursive = T, pattern = ".rda")



  #mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
ensembl <- useEnsembl(biomart = "ensembl",
                     dataset = "hsapiens_gene_ensembl",mirror = "www")

mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 values = results_eQTL$Gene_2,
                 mart = ensembl)

results_eQTL <- results_eQTL %>% 
  inner_join(mapping,by = c("Gene_2" = "ensembl_gene_id")) %>%
  rename(HGNC = hgnc_symbol) %>%
  relocate(HGNC, .after = Gene_2)
  
  # SHowing the whole table
datatable(results_eQTL, rownames = FALSE,
          options = list(scrollX = TRUE),
           class = 'white-space: nowrap')
  



results_psychEncode <- merge_coloc_summaries("/path/to/progtoDementia_psychEncode",
                                 add_signif_SNP = F, recursive = T, pattern = ".rda")

  #mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl",mirror = "www")
mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 values = results_psychEncode$Gene_2,
                 mart = ensembl)

results_psychEncode <- results_psychEncode %>% 
  inner_join(mapping,by = c("Gene_2" = "ensembl_gene_id")) %>%
  rename(HGNC = hgnc_symbol) %>%
  relocate(HGNC, .after = Gene_2)
  

# Showing the results with PP4 > 0.9 
results_psychEncode %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```


## Is there any region with an H4 pp > 0.9 when doing coloc between eQTLgen and GWAS?

```{r}
  # Showing the results with PP4 > 0.9
results_eQTL %>% dplyr::filter(PP.H4.abf > 0.9) %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```


## Is there any region with an H4 pp > 0.9 when doing coloc between psychEncode and GWAS?

```{r}
# Showing the results with PP4 > 0.9 
results_psychEncode %>% dplyr::filter(PP.H4.abf > 0.9) %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')
```


# Session Info
```{r}
utils::sessionInfo()
```
