---
title: "gtexV8_LRP1B"
author: "AMCalejandro"
date: "24/2/2022"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Description.

Transcriptome analyses have previously revealed complexity in regulation of alternative splicing (AS), an event normally controlled by genetic variants. In this new colocalization analysis, we aim to identify whether variants within LRP1B can be interpreted as cis-acting splicing quantitative trait loci (sQTL). To do so, we are going to make use of sQTL data from GTEX_V8.

# Loading libraries

```{r results="hide", warning=FALSE, message=FALSE}
library(tidyverse)
library(data.table)
library(colochelpR)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(SNPlocs.Hsapiens.dbSNP144.GRCh38)
library(DT)
library(coloc)
library(arrow)
library(biomaRt)
```

# Data loading in memory

Since we are focused on LRP1B, I will just explore chr2 gtex_v8_sQTL_Cortex EUR ancestry data.
GTEX_V8 data is in parquet format, easily readable from C++ based arrow lib. 

```{r}
# We load the previously harmonised GWAS df 
# For more information on the processing steps, please ask

survDementia = fread("/path_to/harmonised_progtoDemGWAS_V3.txt")

# Parquet format gtex cortex chrom2
gtex_cortex_chr2 = arrow::read_parquet("/path_to/GTEX_V8/gtexv8_download/GTEx_Analysis_v8_QTLs/GTEx_Analysis_v8_EUR_sQTL_all_associations/Brain_Cortex.v8.EUR.sqtl_allpairs.chr2.parquet")
```


# GTEX / dementia GWAS wrangling

```{r eval = FALSE}
# Minor modification over GWAS df. We use CHR:POS as the SNP column
# We need to get SNP var in hg38 build.
dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh38
survDementia = survDementia %>% dplyr::filter(str_detect(SNP, "^rs")) %>%
  colochelpR::convert_rs_to_loc(SNP_column="SNP", dbSNP= dbSNP) %>%
  dplyr::rename(SNP_RSID = SNP, SNP = loc) %>%
  dplyr::select(-chrpos) %>% # Removing CHR:POS in hg19 buils
  distinct(SNP, .keep_all=T) # Removing extra dups arising.


# GTEX

gtex_cortex_chr2_processed <- gtex_cortex_chr2  %>%
  # Using capturing groups ((...)). Within the last group, I capture the string
  #that does not have ":" till the end
  extract(phenotype_id, into = c('INTRON', 'GENE'), '(.*):+([^:]+)$') %>%
  # Using look ahead to match from ENSG up to the dot
  dplyr::mutate(GENE = stringr::str_extract(GENE, pattern="ENSG.*(?=\\.)")) %>%
  mutate_at("variant_id", ~gsub("chr","", .)) %>%
  separate(variant_id, into = c("CHR", "BP", "REF", "ALT", "BUILD")) %>%
  unite("SNP", CHR:BP, sep = ":", remove = F) %>%
  dplyr::rename(p.value = pval_nominal)
  

# We perform some last minor changes and we finally get the data for LRP1B
gtex_cortex_chr2_processed <- gtex_cortex_chr2_processed %>%
  filter((!is.na(p.value)) & (GENE == "ENSG00000168702")) %>%
  mutate(eQTL = "GTEXV8_CORTEX")

# We also make sure there are no duplicates at the intronic level
# and we check the data is okey for coloc
gtex_cortex_chr2_processed_nodups = gtex_cortex_chr2_processed %>%
  group_by(INTRON) %>% 
  distinct(SNP, .keep_all=T) %>% ungroup() %>%
  check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE)
  
write.table(gtex_cortex_chr2_processed_nodups, "/path_to/gtex_proccessed_CORTEX_chr2.noDUPs_COLOCformatOK.txt", quote = F, sep = "\t", row.names = F, col.names = T) 
```

```{r echo=FALSE}
gtex_cortex_chr2_processed_nodups = fread("/path_to/GTEX_V8/sQTL_EUR_CORTEX/gtex_proccessed_CORTEX_chr2.noDUPs_COLOCformatOK.txt")
```

```{r results="hide", warning = FALSE, message = FALSE, eval = FALSE}
results_path = "/path_to/RESULTS_sQTL_GTEX_CORTEX"
results_path_GWAS_sQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "progtoDementia_sQTL_GTEX_CORTEX_CHR2_LRP1B")
gwas_path = "/path_to/harmonised_progtoDemGWAS_V3.txt"
gtex_cortex_path <- "/path_to/gtex_proccessed_CORTEX_chr2.noDUPs_COLOCformatOK.txt"

# GWAS N
df1_N = as.numeric(unique(survDementia$TotalSampleSize))


intron_vector = unique(gtex_cortex_chr2_processed_nodups$INTRON)

for (index in seq_along(intron_vector)) {
  
  myIntron = intron_vector[index]
  print(str_c(Sys.time(), " - ", index, " - ", myIntron))
  gtex_cortex_filtered <- gtex_cortex_chr2_processed_nodups %>%
    dplyr::filter(INTRON == myIntron) %>%
    dplyr::filter(!duplicated(SNP)) # Cross checking no dups remain
  
  df2_N = round(mean(gtex_cortex_filtered$ma_samples))
  #if (isEmpty(gtex_cortex_filtered$SNP)) {
  #  next
  #}
  # Subset from which I Calculate the proportion cases mean.
  #df_subset <- survDementia_rsID_vbeta %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP) 
  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = survDementia,
                                                           df2 = gtex_cortex_filtered, 
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval", 
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2", 
                                                                     "GENE_2", "INTRON_2"),
                                                        df_1_name = "GWAS", 
                                                        df_2_name = "sQTL",
                                                        df1_path = gwas_path,
                                                        df2_path=gtex_cortex_path,
                                                        p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated, 
                                 results_dir_path=results_path_GWAS_sQTL)
}

```

# Results 

Coloc calculates the posterior probability (PP) for 5 different hypotheses:
- H0: No association with either trait.
- H1: Association with trait 1, not with trait 2.
- H2: Association with trait 2, not with trait 1.
- H3: Association with trait 1 and 2, two independent SNPs.
- H4: Association with trait 1 and trait 2, one shared SNP.

We will set up a PP.H4 > 0.9 to consider the presence of a shared SNP between traits


```{r}
results_sQTL <- merge_coloc_summaries("/path_to/RESULTS_sQTL_GTEX_CORTEX/progtoDementia_sQTL_GTEX_CORTEX_CHR2_LRP1B/",
                                    add_signif_SNP = F, recursive = T, pattern = ".rda")

  # Showing the whole table
datatable(results_sQTL, rownames = FALSE,
          options = list(scrollX = TRUE),
           class = 'white-space: nowrap')
```

