---
title: "gtexV8_LRP1B_APOE"
author: "AMCalejandro"
date: "22/2/2021"
output: html_document
---

# Description.

Transcriptome analysis have previously revealed complexity in regulation of alternative splicing (AS), an event normally controlled by genetic variants. In this new colocalization, we aim to identify whether variants within LRP1B can be interpreted as cis-acting splicing quantitative trait loci (sQTL). To do so, we are going to make use of sQTL data from GTEX_V8.

# Loading libraries

#library(MafDb.1Kgenomes.phase3.hs37d5)
```

```{r results="hide", warning=FALSE, message=FALSE}
.libPaths("/path/to/R_libs/")
library(arrow)
library(colochelpR)
library(data.table)
library(tidyverse)
library(coloc)
library(biomaRt)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(SNPlocs.Hsapiens.dbSNP144.GRCh38)
library(DT)
```



```{r setup, include=FALSE}
theme_rhr <-  theme_bw(base_family = "Helvetica") +
  theme(panel.grid.major.x = element_blank(),
        legend.position = "right",
        strip.text = element_text(size = 8),
        strip.text.y = element_text(angle = 90),
        axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 8),
        axis.title.y = element_text(vjust = 0.6),
        axis.title = element_text(size = 10),
        panel.spacing = unit(0.1, "lines"))
knitr::opts_chunk$set(echo = T, warning = F, message= F)
```

# Data loading in memory
Since we are focused on LRP1B, I will do the analysis over chr2 only. 
GTEX_V8 data is in parquet format, easily readable from C++ based arrow lib. We are just going to explore chr2 gtex_v8_sQTL_Cortex EUR ancestry whole data.

```{r}
# We load the previously harmonised GWAS df 
# For more information on the processing steps, please ask
survDementia = fread("/path/to/harmonised_progtoDemGWAS.txt")
```

```{r eval = FALSE }

gtex_cortex_chr2 = arrow::read_parquet("/path/to/DATA/GTEX_V8/gtexv8_download/GTEx_Analysis_v8_QTLs/GTEx_Analysis_v8_EUR_sQTL_all_associations/Brain_Cortex.v8.EUR.sqtl_allpairs.chr2.parquet")
```


# GTEX wrangling

We are going to use data from cortex.
Again, since all the GTEX_V8 is unprocessed, I will merely load data from chr2 (LRP1B) 
```{r eval = FALSE}
# Minor modification to set SNP columnd to CHR:POS variable
survDementia = survDementia %>% dplyr::rename(SNP_RSID = SNP, SNP = chrpos)

gtex_cortex_chr2 = gtex_cortex_chr2 %>% 
  mutate(phenotype_id = stringr::str_extract(phenotype_id, pattern="ENSG.*(?=\\.)")) %>% # Using look ahead to match from ENSG up to the dot
  mutate_at("variant_id", ~gsub("chr","", .)) %>% 
  separate(variant_id, into = c("CHR", "BP", "REF", "ALT", "BUILD")) %>%
  unite("chrpos", CHR:BP, sep = ":", remove = F)

# Then, the strategy I am going to follow is:
# Group by EMSEMBLID-SNP pair
# Calculate the average Pval
# I decided to set the p.value to pval_nominal.
# Ideally this would be pval_nominal_Average
gtex_cortex_chr2 <- gtex_cortex_chr2 %>% 
  filter(!is.na(pval_nominal)) %>%
  group_by(phenotype_id, chrpos) %>% 
  arrange(pval_nominal) %>% # To make sure I keep the record with the lowest pval_nominal in case whe need it
  mutate(pval_nominal_Average = mean(pval_nominal, na.rm=T)) %>% ungroup() %>%
  distinct(phenotype_id, chrpos, .keep_all=T) %>%
  dplyr::rename(SNP = chrpos, p.value = pval_nominal)

```

```{r echo=FALSE}
gtex_cortex_chr2 = fread("/path/to/DATA/GTEX_V8/sQTL_EUR_CORTEX/gtex_proccessed_chr2_noDUPS_V2.txt")
```

# Getting the genes in +-1MB window

Let's have a look at the coloc PPs for each gene within +-1MB from LRP1B lead SNPs
```{r}
ensembl_gene_ids_overlapping_1Mb_window_hit <- survDementia %>% colochelpR::get_genes_within_1Mb_of_signif_SNPs(pvalue_column="p.value",
                                                                                                                CHR_column="CHR",
                                                                                                                BP_column="BP",
                                                                                                                mart=37)
# Getting the genes present in GTEX
ensembl_gene_ids_overlapping_1Mb_window_hit_sQTLavail = 
  intersect(gtex_cortex_chr2$Gene, ensembl_gene_ids_overlapping_1Mb_window_hit)
```


```{r results="hide", warning = FALSE, message = FALSE, eval = FALSE}
results_path = "/path/to/RESULTS_sQTL_GTEX_CORTEX"
results_path_GWAS_sQTL <- make_results_dir(results_path = results_path,
                                           folder_name = "progtoDementia_sQTL_GTEX_CORTEX_CHR2")
gwas_path = "/path/to/harmonised_progtoDemGWAS_V2.txt"
gtex_cortex_path <- "/path/to/DATA/GTEX_V8/sQTL_EUR_CORTEX/gtex_proccessed_chr2_V2.txt"

# GWAS N
df1_N = as.numeric(unique(survDementia$TotalSampleSize))

# Figurint out GTEX_V8 N
gtex_cortex_covars = fread("/path/to/DATA/GTEX_V8/gtexv8_download/GTEx_Analysis_v8_QTLs/GTEx_Analysis_v8_EUR_sQTL_covariates/Brain_Cortex.v8.EUR.covariates.txt")


for (index in seq_along(ensembl_gene_ids_overlapping_1Mb_window_hit_sQTLavail)) {
  ensembl_geneID <- ensembl_gene_ids_overlapping_1Mb_window_hit_sQTLavail[index]
  print(str_c(Sys.time(), " - ", index, " - ", ensembl_geneID))
  gtex_cortex_chr2_filtered <- gtex_cortex_chr2 %>%
    dplyr::filter(Gene == ensembl_geneID) %>%
    #check_coloc_data_format(beta_or_pval = "pval", check_maf = TRUE) %>%
    dplyr::filter(!duplicated(SNP))
  
  df2_N = round(mean(gtex_cortex_chr2_filtered$ma_samples))
  if (isEmpty(gtex_cortex_chr2_filtered$SNP)) {
    next
  }
  # Subset from which I Calculate the proportion cases mean.
  #df_subset <- survDementia_rsID_vbeta %>% dplyr::filter(SNP %in% ciseQTL_filtered$SNP) 
  coloc_results_annotated <- colochelpR::get_coloc_results(df1 = survDementia,
                                                           df2 = gtex_cortex_chr2_filtered, 
                                                           df1_type = "quant",
                                                           df2_type="quant",
                                                           df1_beta_or_pval="beta",
                                                           df2_beta_or_pval="pval", 
                                                           df1_N = df1_N,
                                                           df2_N = df2_N,
                                                        annotate_signif_SNP_df1_df2 = T,
                                                        key_cols = c("GWAS_1", "eQTL_2", 
                                                                     "Gene_2"),
                                                        df_1_name = "GWAS", 
                                                        df_2_name = "sQTL",
                                                        df1_path = gwas_path,
                                                        df2_path=gtex_cortex_path,
                                                        p12 = 1e-5)
  colochelpR::save_coloc_results(coloc_results_annotated, 
                                 results_dir_path=results_path_GWAS_sQTL)
}

```




# Results 

Coloc calculates the posterior probability (PP) for 5 different hypotheses:
- H0: No association with either trait.
- H1: Association with trait 1, not with trait 2.
- H2: Association with trait 2, not with trait 1.
- H3: Association with trait 1 and 2, two independent SNPs.
- H4: Association with trait 1 and trait 2, one shared SNP.
We will set up a PP.H4 > 0.9 to consider the presence of a shared SNP between traits


```{r}
results_sQTL <- merge_coloc_summaries("/path/to/RESULTS_sQTL_GTEX_CORTEX/progtoDementia_sQTL_GTEX_CORTEX_CHR3.CHR19/",
                                    add_signif_SNP = F, recursive = T, pattern = ".rda")

  #mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
ensembl <- useEnsembl(biomart = "ensembl",
                     dataset = "hsapiens_gene_ensembl",mirror = "www")

mapping <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                 filters = "ensembl_gene_id",
                 values = results_sQTL$Gene_2,
                 mart = ensembl)

results_sQTL <- results_sQTL %>% 
  inner_join(mapping,by = c("Gene_2" = "ensembl_gene_id")) %>%
  dplyr::rename(HGNC = hgnc_symbol) %>%
  relocate(HGNC, .after = Gene_2)
  
  # SHowing the whole table
datatable(results_sQTL, rownames = FALSE,
          options = list(scrollX = TRUE),
           class = 'white-space: nowrap')
```

